File: backend/app/models/events.py
from typing import Literal, Union

from pydantic import BaseModel, Field


class StartSessionEvent(BaseModel):

    type: Literal["session.start"] = "session.start"


class StopSessionEvent(BaseModel):

    type: Literal["session.stop"] = "session.stop"


class AudioInputEvent(BaseModel):

    type: Literal["audio.input"] = "audio.input"
    audio: str = Field(..., description="Base64 encoded PCM16 audio data")


class NavigateEvent(BaseModel):

    type: Literal["slide.navigate"] = "slide.navigate"
    direction: Literal["next", "prev"]


class GoToSlideEvent(BaseModel):

    type: Literal["slide.goto"] = "slide.goto"
    slide_id: int = Field(..., ge=1)


ClientEvent = Union[
    StartSessionEvent,
    StopSessionEvent,
    AudioInputEvent,
    NavigateEvent,
    GoToSlideEvent,
]


class SessionStartedEvent(BaseModel):

    type: Literal["session.started"] = "session.started"
    session_id: str


class SessionStoppedEvent(BaseModel):

    type: Literal["session.stopped"] = "session.stopped"


class AudioOutputEvent(BaseModel):

    type: Literal["audio.output"] = "audio.output"
    audio: str = Field(..., description="Base64 encoded PCM16 audio data")


class AudioDoneEvent(BaseModel):

    type: Literal["audio.done"] = "audio.done"


class SlideChangedEvent(BaseModel):

    type: Literal["slide.changed"] = "slide.changed"
    slide_id: int
    title: str
    content: list[str]
    narration: str
    total_slides: int
    has_next: bool
    has_previous: bool


class TranscriptEvent(BaseModel):

    type: Literal["transcript"] = "transcript"
    text: str
    is_final: bool = False
    speaker: Literal["user", "ai"]


class ErrorEvent(BaseModel):

    type: Literal["error"] = "error"
    message: str
    code: str | None = None


class ConnectionStatusEvent(BaseModel):

    type: Literal["connection.status"] = "connection.status"
    status: Literal["connecting", "connected", "disconnected", "error"]
    message: str | None = None


# Union type for all server events
ServerEvent = Union[
    SessionStartedEvent,
    SessionStoppedEvent,
    AudioOutputEvent,
    AudioDoneEvent,
    SlideChangedEvent,
    TranscriptEvent,
    ErrorEvent,
    ConnectionStatusEvent,
]

File: frontend/src/types/events.ts
export interface StartSessionEvent {
  type: "session.start";
}

export interface StopSessionEvent {
  type: "session.stop";
}

export interface AudioInputEvent {
  type: "audio.input";
  audio: string; // Base64 encoded PCM16
}

export interface NavigateEvent {
  type: "slide.navigate";
  direction: "next" | "prev";
}

export interface GoToSlideEvent {
  type: "slide.goto";
  slide_id: number;
}

export interface CancelResponseEvent {
  type: "response.cancel";
}

export type ClientEvent =
  | StartSessionEvent
  | StopSessionEvent
  | AudioInputEvent
  | NavigateEvent
  | GoToSlideEvent
  | CancelResponseEvent;

export interface SessionStartedEvent {
  type: "session.started";
  session_id: string;
}

export interface SessionStoppedEvent {
  type: "session.stopped";
}

export interface AudioOutputEvent {
  type: "audio.output";
  audio: string; // Base64 encoded PCM16
}

export interface AudioDoneEvent {
  type: "audio.done";
}

export interface SlideChangedEvent {
  type: "slide.changed";
  slide_id: number;
  title: string;
  content: string[];
  narration: string;
  total_slides: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface TranscriptEvent {
  type: "transcript";
  text: string;
  is_final: boolean;
  speaker: "user" | "ai";
}

export interface ErrorEvent {
  type: "error";
  message: string;
  code?: string;
}

export interface ConnectionStatusEvent {
  type: "connection.status";
  status: "connecting" | "connected" | "disconnected" | "error";
  message?: string;
}

export type ServerEvent =
  | SessionStartedEvent
  | SessionStoppedEvent
  | AudioOutputEvent
  | AudioDoneEvent
  | SlideChangedEvent
  | TranscriptEvent
  | ErrorEvent
  | ConnectionStatusEvent;
